File: msad2022_pri/Video.cpp
/*
    Video.cpp
    Copyright © 2022 MSAD Mode2P. All rights reserved.
*/
#include "Video.hpp"
#include "appusr.hpp"

Video::Video() {
#if defined(WITH_OPENCV)
  utils::logging::setLogLevel(utils::logging::LOG_LEVEL_INFO);
  /* set number of threads */
  setNumThreads(0);
  /* prepare the camera */
  cap = VideoCapture(0);
  cap.set(CAP_PROP_FRAME_WIDTH,IN_FRAME_WIDTH);
  cap.set(CAP_PROP_FRAME_HEIGHT,IN_FRAME_HEIGHT);
  cap.set(CAP_PROP_FPS,90);
  assert(cap.isOpened());
  
  /* initial region of interest */
  roi = Rect(0, 0, FRAME_WIDTH, FRAME_HEIGHT);
  /* prepare and keep kernel for morphology */
  kernel = Mat::zeros(Size(7,7), CV_8UC1);
  //kernel = Mat(Size(7,7), CV_8UC1, Scalar(0));
#endif

#if !defined(WITHOUT_X11)
  //XInitThreads();
  disp = XOpenDisplay(NULL);
  sc = DefaultScreenOfDisplay(disp);
  vis = DefaultVisualOfScreen(sc);
  assert(DefaultDepthOfScreen(sc) == 24);
  
  unsigned long black=BlackPixel(disp, 0);
  unsigned long white=WhitePixel(disp, 0);
  win = XCreateSimpleWindow(disp, RootWindow(disp,0), 0, 0, FRAME_WIDTH, 2*FRAME_HEIGHT, 1, black, white);

  XSetWindowAttributes attr;
  attr.override_redirect = True;
  XChangeWindowAttributes(disp, win, CWOverrideRedirect, &attr);
  XMapWindow(disp, win);
  font = XLoadFont(disp, "a14");

  gc = XCreateGC(disp, win, 0, 0);
  XSetForeground(disp, gc, white);
  XSetFont(disp, gc, font);
#endif

  gbuf = malloc(sizeof(unsigned long) * FRAME_WIDTH * 2*FRAME_HEIGHT);
  /* initialize gbuf with zero */
  for (int j = 0; j < 2*FRAME_HEIGHT; j++) {
    for (int i = 0; i < FRAME_WIDTH; i++) {
      buf = (unsigned long*)gbuf + i + j*FRAME_WIDTH;
      *buf = 0;
    }
  }

#if !defined(WITHOUT_X11)
  ximg = XCreateImage(disp, vis, 24, ZPixmap, 0, (char*)gbuf, FRAME_WIDTH, 2*FRAME_HEIGHT, BitmapUnit(disp), 0);
  XInitImage(ximg);
#endif

  /* initial trace target */
  mx = (int)(FRAME_WIDTH/2);
  /* default values */
  gsmin = 0;
  gsmax = 100;
  side = 0;
  rangeOfEdges = 0;
}

Video::~Video() {
#if defined(WITH_OPENCV)
  cap.release();
#endif

#if !defined(WITHOUT_X11)
  XDestroyImage(ximg);
  XFreeGC(disp, gc);
  XDestroyWindow(disp, win);
#endif
}

Mat Video::readFrame() {
  Mat f;
#if defined(WITH_OPENCV)
  cap.read(f);
  /* resize the image for OpenCV processing if exists, otherwise use the previous image */
  if (!f.empty()) {
    if (FRAME_WIDTH != IN_FRAME_WIDTH || FRAME_HEIGHT != IN_FRAME_HEIGHT) {
      Mat img_resized;
      resize(f, img_resized, Size(), (double)FRAME_WIDTH/IN_FRAME_WIDTH, (double)FRAME_HEIGHT/IN_FRAME_HEIGHT);
      assert(img_resized.size().width == FRAME_WIDTH);
      assert(img_resized.size().height == FRAME_HEIGHT);
      f = img_resized;
    }
    frame_prev = f; /* keep the image in case capture() fails to capture a new frame */
  } else {
    f = frame_prev;
  }
#else
  f = nullptr;
#endif
  return f;
}

void Video::writeFrame(Mat f) {
#if defined(WITH_OPENCV)
  if (f.empty()) return;

  if (f.size().width != FRAME_WIDTH || f.size().height != FRAME_HEIGHT) {
    _log("frame size mismatch, w = %d(should be %d), h = %d(should be %d)", f.size().width, FRAME_WIDTH, f.size().height, FRAME_HEIGHT);
    return;
  }

  for (int j = 0; j < FRAME_HEIGHT; j++) {
    for (int i = 0; i < FRAME_WIDTH; i++) {
      buf = (unsigned long*)gbuf + i + j*FRAME_WIDTH;
      int imat = j*f.step + i*f.elemSize();
      *buf = ((f.data[imat+2] << 16) |
	      (f.data[imat+1] << 8) |
	      f.data[imat] );
    }
  }
#elif !defined(WITHOUT_X11)
  const char* MSG = "No OpenCV"; 
  XDrawString(disp, win, gc, 10, 10, MSG, strlen(MSG));
#endif
}

Mat Video::calculateTarget(Mat f) {
#if defined(WITH_OPENCV)
  if (f.empty()) return f;
  
  Mat img_gray, img_bin, img_bin_mor, img_cnt_gray, scan_line;

  /* convert the image from BGR to grayscale */
  cvtColor(f, img_gray, COLOR_BGR2GRAY);
  /* mask the upper half of the grayscale image */
  //for (int i = 0; i < (int)(FRAME_HEIGHT/2); i++) {
  //  for (int j = 0; j < FRAME_WIDTH; j++) {
  //    img_gray.at<uchar>(i,j) = 255; /* type = CV_8U */
  //  }
  //}
  /* binarize the image */
  inRange(img_gray, gsmin, gsmax, img_bin);
  /* remove noise */
  morphologyEx(img_bin, img_bin_mor, MORPH_CLOSE, kernel);

  /* focus on the region of interest */
  Mat img_roi(img_bin_mor, roi);
  /* find contours in the roi with offset */
  vector<vector<Point>> contours;
  vector<Vec4i> hierarchy;
  findContours(img_roi, contours, hierarchy, RETR_EXTERNAL, CHAIN_APPROX_SIMPLE, Point(roi.x,roi.y));
  /* identify the largest contour */
  if (contours.size() >= 1) {
    int i_target = 0;
    if (contours.size() >= 2) {
      vector<vector<int>> contAreas;
      for (int i = 0; i < (int)contours.size(); i++) {
	vector<int> contArea;
	contArea.push_back((int)contourArea(contours[i])); /* first element - area of the contour */
	contArea.push_back(i); /* second element - index to contours */
	contAreas.push_back(contArea);
      }
      /* sort contAreas in descending order by area size */
      sort(contAreas.begin(),contAreas.end(),[](const vector<int> &alpha,const vector<int> &beta){return alpha[0] > beta[0];});
      /* calculate the bounding box around the two largest contours,
         either of which is to be set as the new region of interest */
      Rect bb1st = boundingRect(contours[contAreas[0][1]]);
      Rect bb2nd = boundingRect(contours[contAreas[1][1]]);
      if ( (bb2nd.y + bb2nd.height >= roi.height) && /* the second largest contour touches the roi bottom? */
	   (2*contAreas[1][0] >= contAreas[0][0]) ) { /* the second largest contour is large enough? */
	/* yes then compare the x location of 1st and 2nd bounding rectangles */
	if (side == 0) { /* tracing the left edge needs the rectanble on the left */
	  if (bb1st.x <= bb2nd.x) {
	    i_target = contAreas[0][1];
	    roi = bb1st;
	  } else {
	    i_target = contAreas[1][1];
	    roi = bb2nd;
	  }
	} else if (side == 1) { /* tracing the right edge needs the rectangle on the right */
	  if (bb1st.x + bb1st.width >= bb2nd.x + bb2nd.width) {
	    i_target = contAreas[0][1];
	    roi = bb1st;
	  } else {
	    i_target = contAreas[1][1];
	    roi = bb2nd;
	  }
	} else { /* tracing the line center goes after the largest contour */
	  i_target = contAreas[0][1];
	  roi = bb1st;
	}
      } else { /* the second largest contour does not touch the roi bottom and shall be ignored */
	i_target = contAreas[0][1];
	roi = bb1st;
      }
    }
    /* draw the target contour on the original image */
    polylines(f, (vector<vector<Point>>){contours[i_target]}, 0, Scalar(0,255,0), LINE_THICKNESS);

    /* adjust the region of interest */
    roi.x = roi.x - ROI_BOUNDARY;
    roi.y = roi.y - ROI_BOUNDARY;
    roi.width = roi.width + 2*ROI_BOUNDARY;
    roi.height = roi.height + 2*ROI_BOUNDARY;
    if (roi.x < 0) {
      roi.x = 0;
    }
    if (roi.y < 0) {
      roi.y = 0;
    }
    if (roi.x + roi.width > FRAME_WIDTH) {
      roi.width = FRAME_WIDTH - roi.x;
    }
    if (roi.y + roi.height > FRAME_HEIGHT) {
      roi.height = FRAME_HEIGHT - roi.y;
    }
 
    /* prepare for trace target calculation */
    /*
      Note: Mat::zeros with CV_8UC3 does NOT work and don't know why
    */
    Mat img_cnt(f.size(), CV_8UC3, Scalar(0,0,0));
    drawContours(img_cnt, (vector<vector<Point>>){contours[i_target]}, 0, Scalar(0,255,0), 1);
    cvtColor(img_cnt, img_cnt_gray, COLOR_BGR2GRAY);
    /* scan the line really close to the image bottom to find edges */
    scan_line = img_cnt_gray.row(img_cnt_gray.size().height - LINE_THICKNESS);
    /* convert the Mat to a NumCpp array */
    auto scan_line_nc = nc::NdArray<nc::uint8>(scan_line.data, scan_line.rows, scan_line.cols);
    auto edges = scan_line_nc.flatnonzero();
    if (edges.size() >= 2) {
      rangeOfEdges = edges[edges.size()-1] - edges[0];
      if (side == 0) {
	mx = edges[0];
      } else if (side == 1) {
	mx = edges[edges.size()-1];
      } else {
	mx = (int)((edges[0]+edges[edges.size()-1]) / 2);
      }
    } else if (edges.size() == 1) {
      rangeOfEdges = 1;
      mx = edges[0];
    }
  } else { /* contours.size() == 0 */
    rangeOfEdges = 0;
    roi = Rect(0, 0, FRAME_WIDTH, FRAME_HEIGHT);
  }
  //_logNoAsp("roe = %d", rangeOfEdges);

  /* draw the area of interest on the original image */
  rectangle(f, Point(roi.x,roi.y), Point(roi.x+roi.width,roi.y+roi.height), Scalar(255,0,0), LINE_THICKNESS);
  /* draw the trace target on the image */
  circle(f, Point(mx, FRAME_HEIGHT-LINE_THICKNESS), CIRCLE_RADIUS, Scalar(0,0,255), -1);
#endif /* defined(WITH_OPENCV) */

  /* calculate variance of mx from the center in pixel */
  int vxp = mx - (int)(FRAME_WIDTH/2);
  /* convert the variance from pixel to milimeters
     72 is length of the closest horizontal line on ground within the camera vision */
  float vxm = vxp * 72 / FRAME_WIDTH;
  /* calculate the rotation in degree (z-axis)
     284 is distance from axle to the closest horizontal line on ground the camera can see */
  if (vxm == 0) {
    theta = 0;
  } else {
    theta = 180 * atan(vxm / 284) / M_PI;
  }
  //_logNoAsp("mx = %d, vxm = %d, theta = %d", mx, (int)vxm, (int)theta);

  return f;
}

void Video::show() {
  sprintf(strbuf[0], "x=%+05d,y=%+05d", plotter->getLocX(), plotter->getLocY());
  sprintf(strbuf[1], "dist=%06d", plotter->getDistance());
  sprintf(strbuf[2], "deg=%03d,gyro=%+04d", plotter->getDegree(), gyroSensor->getAngle());
  sprintf(strbuf[3], "pwR=%+04d,pwL=%+04d", rightMotor->getPWM(), leftMotor->getPWM());

#if !defined(WITHOUT_X11)
  XPutImage(disp, win, gc, ximg, 0, 0, 0, 0, FRAME_WIDTH, 2*FRAME_HEIGHT);
  XDrawString(disp, win, gc, DATA_INDENT, FRAME_HEIGHT+2*DATA_INDENT, strbuf[0], strlen(strbuf[0]));
  XDrawString(disp, win, gc, DATA_INDENT, FRAME_HEIGHT+5*DATA_INDENT, strbuf[1], strlen(strbuf[1]));
  XDrawString(disp, win, gc, DATA_INDENT, FRAME_HEIGHT+8*DATA_INDENT, strbuf[2], strlen(strbuf[2]));
  XDrawString(disp, win, gc, DATA_INDENT, FRAME_HEIGHT+11*DATA_INDENT, strbuf[3], strlen(strbuf[3]));
  XFlush(disp);
#endif
}

float Video::getTheta() {
  return theta;
}

int Video::getRangeOfEdges() {
  return rangeOfEdges;
}

void Video::setThresholds(int gsMin, int gsMax) {
  gsmin = gsMin;
  gsmax = gsMax;
}

void Video::setTraceSide(int traceSide) {
  side = traceSide;
}

File: 2024base/py_etrobo_util/video.py
import cv2
import math
import numpy as np
from enum import Enum
from picamera2 import Picamera2, Preview
from .plotter import Plotter
from etrobo_python import Hub, Motor, ColorSensor, TouchSensor, SonarSensor, GyroSensor

def round_up_to_odd(f) -> int:
    return int(np.ceil(f / 2.) * 2 + 1)

# frame size for Raspberry Pi camera capture
IN_FRAME_WIDTH  = 640
IN_FRAME_HEIGHT = 480

# frame size for OpenCV
#FRAME_WIDTH  = 160
#FRAME_HEIGHT = 120
FRAME_WIDTH  = 320
FRAME_HEIGHT = 240

CROP_WIDTH     = int(13*FRAME_WIDTH/16) # for full angle
#CROP_WIDTH     = int(9*FRAME_WIDTH/16)
CROP_HEIGHT    = int(3*FRAME_HEIGHT/8) # for full angle
#CROP_HEIGHT    = int(2*FRAME_HEIGHT/8)
CROP_U_LIMIT   = FRAME_HEIGHT-CROP_HEIGHT
CROP_D_LIMIT   = FRAME_HEIGHT
CROP_L_LIMIT   = int((FRAME_WIDTH-CROP_WIDTH)/2)
CROP_R_LIMIT   = (CROP_L_LIMIT+CROP_WIDTH)
MORPH_KERNEL_SIZE = round_up_to_odd(int(FRAME_WIDTH/48))
ROI_BOUNDARY   = int(FRAME_WIDTH/10)
LINE_THICKNESS = int(FRAME_WIDTH/80)
CIRCLE_RADIUS  = int(FRAME_WIDTH/40)
SCAN_V_POS     = int(13*FRAME_HEIGHT/16 - LINE_THICKNESS) # for full angle
#SCAN_V_POS     = int(16*FRAME_HEIGHT/16 - LINE_THICKNESS)

# frame size for X11 painting
OUT_FRAME_WIDTH  = 160
OUT_FRAME_HEIGHT = 120


class TraceSide(Enum):
    NORMAL = "Normal"
    OPPOSITE = "Opposite"
    RIGHT = "Right"
    LEFT = "Left"
    CENTER = "Center"


class Video(object):
    def __init__(self):
        cv2.setLogLevel(3) # LOG_LEVEL_WARNING
        # set number of threads
        #cv2.setNumThreads(0)
        # prepare the camera
        self.pc2 = Picamera2()
        full_reso = self.pc2.camera_properties['PixelArraySize']
        config = self.pc2.create_preview_configuration(main={"format": 'RGB888', "size": (IN_FRAME_WIDTH, IN_FRAME_HEIGHT)}, raw={"size": full_reso})
        self.pc2.configure(config)
        self.pc2.start()

        # initial region of interest
        self.roi = (CROP_L_LIMIT, CROP_U_LIMIT, CROP_WIDTH, CROP_HEIGHT)
        # prepare and keep kernel for morphology
        self.kernel = np.ones((MORPH_KERNEL_SIZE,MORPH_KERNEL_SIZE), np.uint8)
        # initial trace target
        self.cx = int(FRAME_WIDTH/2)
        self.cy = SCAN_V_POS
        self.mx = self.cx
        # default values
        self.gsmin = 0
        self.gsmax = 100
        self.trace_side = TraceSide.NORMAL
        self.theta:float = 0
        self.target_insight = False

    def __del__(self):
        cv2.destroyAllWindows
        self.pc2.stop()
        del pc2


    def process(self,
                plotter: Plotter,
                hub: Hub,
                arm_motor: Motor,
                right_motor: Motor,
                left_motor: Motor,
                touch_sensor: TouchSensor,
                color_sensor: ColorSensor,
                sonar_sensor: SonarSensor,
                gyro_sensor: GyroSensor,
                ) -> None:
        frame = self.pc2.capture_array()

        # clone the image if exists, otherwise use the previous image
        if len(frame) != 0:
            img_orig = frame.copy()
            # resize the image for OpenCV processing
        if FRAME_WIDTH != IN_FRAME_WIDTH or FRAME_HEIGHT != IN_FRAME_HEIGHT:
            img_orig = cv2.resize(img_orig, (FRAME_WIDTH,FRAME_HEIGHT))
            if img_orig.shape[1] != FRAME_WIDTH or img_orig.shape[0] != FRAME_HEIGHT:
                sys.exit(-1)
        # convert the image from BGR to grayscale
        img_gray = cv2.cvtColor(img_orig, cv2.COLOR_BGR2GRAY)
        # crop a part of image for binarization
        img_gray_part = img_gray[CROP_U_LIMIT:CROP_D_LIMIT, CROP_L_LIMIT:CROP_R_LIMIT]
        # binarize the image
        img_bin_part = cv2.inRange(img_gray_part, self.gsmin, self.gsmax)
        # prepare an empty matrix
        img_bin = np.zeros((FRAME_HEIGHT, FRAME_WIDTH), np.uint8)
        # copy img_bin_part into img_bin
        img_bin[CROP_U_LIMIT:CROP_U_LIMIT+img_bin_part.shape[0], CROP_L_LIMIT:CROP_L_LIMIT+img_bin_part.shape[1]] = img_bin_part
        # remove noise
        img_bin_mor = cv2.morphologyEx(img_bin, cv2.MORPH_CLOSE, self.kernel)
        # convert the binary image from grayscale to BGR for later
        img_bin_rgb = cv2.cvtColor(img_bin_mor, cv2.COLOR_GRAY2BGR)

        # focus on the region of interest
        x, y, w, h = self.roi
        img_roi = img_bin_mor[y:y+h, x:x+w]
        # find contours in the roi with offset
        contours, hierarchy = cv2.findContours(img_roi, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE, offset=(x,y))
        # identify the largest contour
        if len(contours) >= 1:
            i_area_max = 0
            area_max = 0
            for i, cnt in enumerate(contours):
                area = cv2.contourArea(cnt)
                if area > area_max:
                    area_max = area
                    i_area_max = i
            # draw the largest contour on the original image
            #img_orig = cv2.drawContours(img_orig, [contours[i_area_max]], 0, (0,255,0), LINE_THICKNESS)
            img_orig = cv2.polylines(img_orig, [contours[i_area_max]], 0, (0,255,0), LINE_THICKNESS)

            # calculate the bounding box around the largest contour
            # and set it as the new region of interest
            x, y, w, h = cv2.boundingRect(contours[i_area_max])
            # adjust the region of interest
            x = x - ROI_BOUNDARY
            y = y - ROI_BOUNDARY
            w = w + 2*ROI_BOUNDARY
            h = h + 2*ROI_BOUNDARY
            if x < CROP_L_LIMIT:
                x = CROP_L_LIMIT
            if y < CROP_U_LIMIT:
                y = CROP_U_LIMIT
            if x + w > CROP_R_LIMIT:
                w = CROP_R_LIMIT - x
            if y + h > CROP_D_LIMIT:
                h = CROP_D_LIMIT - y
            # set the new region of interest
            self.roi = (x, y, w, h)
        
            # prepare for trace target calculation
            img_cnt = np.zeros_like(img_orig)
            img_cnt = cv2.drawContours(img_cnt, [contours[i_area_max]], 0, (0,255,0), 1)
            img_cnt_gray = cv2.cvtColor(img_cnt, cv2.COLOR_BGR2GRAY)
            # scan the line at SCAN_V_POS to find edges
            scan_line = img_cnt_gray[SCAN_V_POS]
            edges = np.flatnonzero(scan_line)
            # calculate the trace target using the edges
            if len(edges) >= 2:
                if self.trace_side == TraceSide.LEFT:
                    self.cx = edges[0]
                elif self.trace_side == TraceSide.RIGHT:
                    self.cx = edges[len(edges)-1]
                else:
                    self.cx = int((edges[0]+edges[len(edges)-1]) / 2)
            elif len(edges) == 1:
                self.cx = edges[0]
            self.mx = self.cx
            self.target_insight = True
        else: # len(contours) == 0
            self.roi = (CROP_L_LIMIT, CROP_U_LIMIT, CROP_WIDTH, CROP_HEIGHT)
            # keep mx in order to maintain the current move of robot
            self.cx = int(FRAME_WIDTH/2)
            self.cy = SCAN_V_POS
            self.target_insight = False
            

        # draw the area of interest on the original image
        x, y, w, h = self.roi
        if self.target_insight:
            cv2.rectangle(img_orig, (x,y), (x+w,y+h), (255,0,0), LINE_THICKNESS)
        else:
            cv2.rectangle(img_orig, (x,y), (x+w,y+h), (0,0,255), LINE_THICKNESS)
        # draw the trace target on the image
        cv2.circle(img_orig, (self.mx, SCAN_V_POS), CIRCLE_RADIUS, (0,0,255), -1)
        # calculate variance of mx from the center in pixel
        vxp = self.mx - int(FRAME_WIDTH/2)
        # convert the variance from pixel to milimeters
        # 138 is length of the closest horizontal line on ground within the camera vision
        vxm = vxp * 138 / FRAME_WIDTH
        # calculate the rotation in radians (z-axis)
        # 205 is distance from axle to the closest horizontal line on ground the camera can see
        self.theta = 180 * math.atan(vxm / 205) / math.pi
        #print(f"mx = {self.mx}, vxm = {vxm}, theta = {self.theta}")

        # prepare text area
        img_text = np.zeros((FRAME_HEIGHT, FRAME_WIDTH, 3), np.uint8)
        if not plotter is None:
            try:
                cv2.putText(img_text, f"ODO={plotter.get_distance():+06}", (0,15), cv2.FONT_HERSHEY_PLAIN, 1.2, (255,255,255), 1, cv2.LINE_AA)
                cv2.putText(img_text, f"cx={self.cx:} cy={self.cy} T={self.theta:+06.1f}", (0,30), cv2.FONT_HERSHEY_PLAIN, 1.2, (255,255,255), 1, cv2.LINE_AA)
                cv2.putText(img_text, f"mV={hub.get_battery_voltage():04} mA={hub.get_battery_current():04}", (0,45), cv2.FONT_HERSHEY_PLAIN, 1.2, (255,255,255), 1, cv2.LINE_AA)
            except Exception as e:
                pass
        # concatinate the images - original + text area
        img_comm = cv2.vconcat([img_orig,img_text])
        # shrink the image to avoid delay in transmission
        if OUT_FRAME_WIDTH != FRAME_WIDTH or OUT_FRAME_HEIGHT != FRAME_HEIGHT:
            img_comm = cv2.resize(img_comm, (OUT_FRAME_WIDTH,2*OUT_FRAME_HEIGHT))
        # transmit and display the image
        cv2.imshow("video monitor", img_comm)

        c = cv2.waitKey(1) # show the window
        
    def get_theta(self) -> float:
        return self.theta

    def set_thresholds(self, gs_min: int, gs_max: int) -> None:
        self.gs_min = gs_min
        self.gs_max = gs_max

    def set_trace_side(self, trace_side: TraceSide) -> None:
        self.trace_side = trace_side
